lightwood/encoder/datetime/datetime.py

```diff
@@ -1,90 +1,82 @@
-        self.constants = {'year': 3000.0, 'month': 12.0, 'weekday': 7.0,
-                          'hour': 24.0, 'minute': 60.0, 'second': 60.0}
-        self.output_size = 7
+        self.constant_keys = ['year', 'month', 'day', 'hour', 'minute', 'second']
+        self.constant_vals = torch.Tensor([3000.0, 12.0, 31.0, 24.0, 60.0, 60.0])  # cycle length
+        self.constant_map = {k: v.item() for k, v in zip(self.constant_keys, self.constant_vals)}
 
-    def prepare(self, priming_data):
-        if self.is_prepared:
-            raise Exception('You can only call "prepare" once for a given encoder.')
+        self.output_size = len(self.constant_keys)
+        self.empty_vector = np.zeros((self.output_size, ))
 
-        self.is_prepared = True
+        self.max_vals = torch.Tensor([pd.Timestamp.max.year - 1, 12, 31, 23, 59, 59])
+        self.min_vals = torch.Tensor([pd.Timestamp.min.year + 1, 1, 1, 0, 0, 0])
 
-    def encode(self, data):
-        """
-        :param data: # @TODO: receive a consistent data type here; currently either list of lists or pd.Series w/lists
-        :return: encoded data
-        """
-        if not self.is_prepared:
-            raise Exception('You need to call "prepare" before calling "encode" or "decode".')
-
-        ret = [self.encode_one(unix_timestamp) for unix_timestamp in data]
-
-        return torch.Tensor(ret)
+    def prepare(self, priming_data):
+        self.is_prepared = True
 
-    def encode_one(self, unix_timestamp: Optional[float]):
+    def encode(self, data: Union[np.ndarray, pd.Series]) -> torch.Tensor:
         """
-        Encodes a list of unix_timestamps, or a list of tensors with unix_timestamps
-        :param data: list of unix_timestamps (unix_timestamp resolution is seconds)
-        :return: a list of vectors
+        :param data: a pandas series with numerical dtype, previously cleaned with dataprep_ml
+        :return: encoded data, shape (len(data), self.output_size)
         """
-        if is_none(unix_timestamp):
-            vector = [0] * len(self.fields)
-        else:
-            c = self.constants
-            date = datetime.datetime.fromtimestamp(unix_timestamp)
-            day_constant = calendar.monthrange(date.year, date.month)[1]
-            vector = [date.year / c['year'], date.month / c['month'], date.day / day_constant,
-                      date.weekday() / c['weekday'], date.hour / c['hour'],
-                      date.minute / c['minute'], date.second / c['second']]
-        return vector
-
-    def decode(self, encoded_data, return_as_datetime=False):
-        ret = []
+        if type(data) not in (np.ndarray, pd.Series):
+            raise Exception(f'Data should be pd.Series or np.ndarray! Got: {type(data)}')
+        if isinstance(data, np.ndarray):
+            data = pd.Series(data)
+
+        data = data.fillna(pd.Timestamp.max.timestamp())  # TODO: replace with mean?
```

Comment:
I would an actual interpolation step using a linear spline here, as nothing guarantees that the timestamps are evenly sampled (i.e., the frequency of data acquisition is constant). fillna() is somewhat imprecise and I've seen it producing wrong results when I worked with time series data.

Reply:
fillna() can be used to perform interpolation, by the way, but since we do have access to the "raw" unix timestamps there is no need for it. 

Also, please note that in some cases we will actually _want to keep_ the NaNs around to find gaps in the data.

Reply:
Great idea! I'll add it as a separate issue because it seems quite important to handle this in a better way.